### PR TITLE
[RR-159] setup/teardown sessions on the log

### PR DESCRIPTION
**Pull request by sjpotter**
handle watch/unwatch at a basic level to support sessions on the log
    
doesn't do anything with it besides allocating a void * on session creation and free on session deletion.
    
includes metrics to count number of active sessions for testing purposes

### CODE REVIEW

File: src/raft.c [hunk 1]
```diff
@@ -347,6 +348,17 @@ void RaftExecuteCommandArray(RedisRaftCtx *rr,
         size_t cmdlen;
         const char *cmd = RedisModule_StringPtrLen(c->argv[0], &cmdlen);
 
+        if (strncasecmp("watch", cmd, cmdlen) == 0) {
```
Comment (tezc) [inline src/raft.c:351]: `strncasecmp("watch", "w", 1) == 0` will enter this if check. So, I suggest adding len check first:

```c
if (cmdlen == 5 && strncasecmp("watch", cmd, cmdlen) == 0) {
```

File: src/redisraft.c [hunk 2]
```diff
@@ -555,6 +555,30 @@ static void handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
     RedisModule_ReplyWithSimpleString(ctx, "OK");
 }
 
+static void handleEndSession(RedisRaftCtx *rr, RedisModuleCtx *ctx, char *reason)
+{
+    RaftReq *req = RaftReqInit(ctx, RR_END_SESSION);
+    if (req == NULL) {
```
Comment (tezc) [inline src/redisraft.c:561]: It won't be null ever. We can skip the check

File: src/redisraft.c [hunk 3]
```diff
@@ -597,6 +621,11 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
         return true;
     }
 
+    if (len == strlen("UNWATCH") && strncasecmp(cmd_str, "UNWATCH", len) == 0) {
```
Comment (tezc) [inline src/redisraft.c:624]: Why do we handle `unwatch` here but handle `watch` in the state machine? Let's send `unwatch` there. We can handle it next to `watch`.

[the remark this answers is not in the document]
Comment (tezc) [reply src/redisraft.c:624]: also `WATCH` can be inside of multi-exec?. We need to check Redis' behavior and mimic it here.

File: src/log.c [hunk 4]
```diff
@@ -138,6 +138,7 @@ static size_t generateEntryHeader(raft_entry_t *ety, unsigned char *buf, size_t
     pos += multibulkWriteLong(pos, end - pos, ety->id);
     pos += multibulkWriteUInt64(pos, end - pos, ety->session);
     pos += multibulkWriteInt(pos, end - pos, ety->type);
+
```
Comment (tezc) [inline src/log.c:141]:
```suggestion
```
Is this change intentional? If not, let's get rid of it.

File: src/raft.c [hunk 5]
```diff
@@ -347,6 +348,17 @@ void RaftExecuteCommandArray(RedisRaftCtx *rr,
...
+            unsigned long long id = cmds->client_id;
+
+            int nokey;
+            RedisModule_DictGetC(rr->session_dict, &id, sizeof(id), &nokey);
+            if (nokey) {
+                void *session = RedisModule_Alloc(sizeof(void *));
+                RedisModule_DictSetC(rr->session_dict, &id, sizeof(id), session);
```
[the remark this answers is not in the document]
Comment (tezc) [reply src/raft.c:358]: Not sure if this is the same thing but I think, instead of `session` in libraft, we can call it `client_id`.

```c
struct raft_entry_t { 
    ...
    raft_client_id client_id; // Instead of raft_session session
}
```
I think this is easier to understand and also it matches with the function name of redis: `RM_GetClientId()` -> use the return value to set client id in the entry. 

You set the client id of the entry, then do whatever you want to do with it, implement sessions as we do. 

Now it requires libraft change but if we all agree it is better, we can quickly rename it? Please comment.

[the remark this answers is not in the document]
Comment (sjpotter) [reply src/raft.c:358]: so the argument i made before is that libraft shouldn't be tied to redis semantics.  in redis a "session" is tied to a client id, but it doesn't have to be that way in other systems.  I'm not tied to this interpretation, but I feel "session" should be generic, while the implementation on top of libraft should refer to it whatever makes the most sense for it.

I can change the session variable here to be more understandable.

File: src/redisraft.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:624]: WATCH can't be inside a multiexec (tested previously).  Apparently, UNWATCH can be.  However, it seems to always cause the command to return nil.  the redis semantics seem to be poor here.

[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:624]: change that.  nil on dirty is what exec always returns.  so unwatch within a multi/exec dirties the session.  weird.  In this case, I'd think we would have to think to decide what to do, as we aren't (currently) queing the multi/exec on the log, we can know at unwatch time if we are in an multi and just error it out.

File: src/raft.c (hunk 1 above)
Comment (sjpotter) [reply src/raft.c:351]: ok, makes sense.

File: src/redisraft.c (hunk 2 above)
Comment (sjpotter) [reply src/redisraft.c:561]: well, this code is about to be rewritten, req can't be created here (no req on disconnect, this actually was the problem I hit earlier that I thought was a redis bug, still might be, but I was doing something bad.  So will have to have a req ==/!= NULL code path, but this isn't it.

File: src/redisraft.c (hunk 3 above)
Comment (sjpotter) [reply src/redisraft.c:624]: confused what you mean that UNWATCH isn't being handled in the state machine?  this appends a raft entry.  no actually unwatch happens here.  The point of handling it here, is to not handle it a "redis command" and append it as a redis command, but as a unique "end session" object.

File: src/log.c (hunk 4 above)
Comment (sjpotter) [reply src/log.c:141]: it sort of was.  the ety-> entries are the raft_entry_t header (id/session/type), the next entry really belong with the data (i.e. we have data_len here, but data is appended separately) which gets appended externally to this, so wanted to make it clear.

File: src/redisraft.c (hunk 3 above)
Comment (tezc) [reply src/redisraft.c:624]: Currently, in `RaftExecuteCommandArray()` you have

```c
if (strncasecmp("watch", cmd, cmdlen) == 0) {
....
    createSession();
...
} 
```

you can add this part:

```c
if (strncasecmp("watch", cmd, cmdlen) == 0) {
...
    createSession();
...
} else if (strncasecmp("unwatch", cmd, cmdlen) == 0) {
...
    endSession();
...
}
```

Then, we don't need to intercept `unwatch` here (on append time), similar to `watch`. So, we handle these commands specially in the state machine. 

Then, we'll use `RAFT_LOGTYPE_END_SESSION` entry just for disconnect events.

[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:624]: so an even more update.  unwatch is usable in multi (so current code is problematic), BUT, it doesn't do anything in reality.  i.e. if the watched value in concurrently modified, nil is returned (as multi is dirty), which makes sense as UNWATCH only would execute on exec, but already dirty).  And if its not dirty, you just get an OK back.  It would make more sense to reject it, like it rejects WATCH

(error) ERR WATCH inside MULTI is not allowed

File: src/raft.c [hunk 6]
```diff
@@ -518,6 +530,31 @@ static void unlockDeleteKeys(RedisRaftCtx *rr, raft_entry_t *entry)
...
+
+static void endSession(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_END_SESSION);
+    RaftReq *req = entry->user_data;
+
+    unsigned long long id = entry->session;
+
+    void *session = NULL;
+    RedisModule_DictDelC(rr->session_dict, &id, sizeof(id), &session);
```
[the remark this answers is not in the document]
Comment (fadidahanna) [reply src/raft.c:546]: Agree that the implementation of `RedisModule_DictDelC` do it, but we should stick to the spec, and verify also the return value

File: src/raft.c [hunk 7]
```diff
@@ -347,6 +348,17 @@ void RaftExecuteCommandArray(RedisRaftCtx *rr,
         size_t cmdlen;
         const char *cmd = RedisModule_StringPtrLen(c->argv[0], &cmdlen);
 
+        if (strncasecmp("watch", cmd, cmdlen) == 0) {
+            unsigned long long id = cmds->client_id;
+
+            int nokey;
+            RedisModule_DictGetC(rr->session_dict, &id, sizeof(id), &nokey);
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/raft.c:355]: having a getSession() that returns the session object / creates it if neccesary (first watch)

File: src/raft.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/raft.c:546]: I believe this is fine?  session = NULL. after DictDelC, either it will still be NULL (no session set for this id) or it won't be NULL (yes, session).

as we then do

    if (session) {
        freeSession(session);
    }
    
it should all be fine.

File: src/redisraft.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:624]: refactored, so UNWATCH will be handled in normal redis command path so will be enqueued into a MULTI (pointless, but what redis does)

File: src/raft.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/raft.c:358]: session -> client_session and functions are *ClientSession()

File: src/raft.c (hunk 1 above)
Comment (sjpotter) [reply src/raft.c:351]: done

File: src/redisraft.c (hunk 3 above)
Comment (sjpotter) [reply src/redisraft.c:624]: its not just for disconnect, it can also be used for multi abort / discard.

File: src/log.c (hunk 4 above)
Comment (tezc) [reply src/log.c:141]: Change is okay but probably it'd be nicer as part of another PR. As this is the only change in log.c, when you look that this PR later or to the commit history, it'll give you an impression that we changed log.c as part of this PR.

File: src/raft.c [hunk 8]
```diff
@@ -341,6 +366,9 @@ void RaftExecuteCommandArray(RedisRaftCtx *rr,
         return;
     }
 
+    void *client_session = getClientSession(rr, cmds);
+    if (client_session) {} /* avoid unused variable check */
```
Comment (tezc) [inline src/raft.c:370]:
```suggestion
    (void) client_sessions;
```
This should also work, just fyi.

File: src/raft.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (tezc) [reply src/raft.c]: I thought we would change variable name in libraft, no?

To be honest, I don't like session->client_session change. In general I don't like verbose naming as it makes code larger/longer. `session` was perfectly clear to me, "client_session" doesn't make it more readable imho. 

Just to be clear, my point was referring `client_id` <-> `session` <-> `session_id` interchangeably in the code. I find this a bit confusing / harder to understand. I think confusion comes from `raft_entry_t` as it has `session` variable (it is an id) and we also have "session" struct/logic in redisraft. To fix this issue, we can rename `raft_entry_t`'s `session` to `client_id`. Then, it'll be easier to understand, at least this was my confusion / problem.

Anyway, I'm not even sure we discuss same thing :) Maybe I misunderstood initial comment in this thread.

File: src/redisraft.c (hunk 3 above)
Comment (tezc) [reply src/redisraft.c]: We can directly write `UNWATCH`, `DISCARD` commands to the log and process differently (by destroying related session) in the state machine. 

It'll be same functionality-wise but a step closer to the "real sessions". Send commands/events to the state machine and process there.

Then, for the disconnect event, we create that entry(`RAFT_LOGTYPE_CLIENT_DISCONNECT`)  and write to the log. So, these two commands (unwatch, discard) and disconnect event will do the same thing currently: destroy the session. In future, maybe they won't. (depends on what we implement).

Comment (sjpotter) [reply src/redisraft.c]: the reason I didn't do it that way, is I dislike the text parsing of UNWATCH/DISCARD at apply time.  It just bothers me when dealing with the state machine to work based on string comparisons.  if we want to do it that way, I can implement it, just rus me the wrong way.

Comment (tezc) [reply src/redisraft.c]: ignoring for now as this is not critical

File: src/raft.c [hunk 9]
```diff
@@ -889,6 +946,8 @@ static int raftApplyLog(raft_server_t *raft, void *user_data, raft_entry_t *entr
         case RAFT_LOGTYPE_DELETE_UNLOCK_KEYS:
             unlockDeleteKeys(rr, entry);
             break;
+        case RAFT_LOGTYPE_END_SESSION:
+            handleEndClientSession(rr, entry);
```
Comment (tezc) [inline src/raft.c:950]:
```suggestion
            handleEndClientSession(rr, entry);
            break;
```

Let's add this not to see a surprise later

File: src/redisraft.h [hunk 10]
```diff
@@ -418,9 +418,10 @@ typedef struct RedisRaftCtx {
     unsigned long snapshotreq_received;          /* Number of received snapshotreq messages */
     unsigned long exec_throttled;                /* Number of command executions throttled due to slow execution */
 
-    int entered_eval;             /* handling a lua script */
-    RedisModuleDict *locked_keys; /* keys thar have been locked for migration */
-    RedisModuleDict *acl_dict;    /* maps acl strings to RedisModuleUser * objects */
+    int entered_eval;                     /* handling a lua script */
+    RedisModuleDict *locked_keys;         /* keys thar have been locked for migration */
```
Comment (tezc) [inline src/redisraft.h:422]:
```suggestion
    RedisModuleDict *locked_keys;         /* keys that have been locked for migration */
```
let's fix this on the way

File: src/redisraft.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tezc) [reply src/redisraft.c]: Okay, are we going to prevent WATCH inside multi in this PR or in another PR? 
Btw, I guess we can even just return OK to UNWATCH without adding it to the queue.

File: src/redisraft.c [hunk 11]
```diff
@@ -621,6 +663,10 @@ static void handleRedisCommandAppend(RedisRaftCtx *rr,
         return;
     }
 
+    if (handleUnwatch(rr, ctx, cmds)) {
```
Comment (tezc) [inline src/redisraft.c:666]: I'm scared to ask this question but hope ACL does not apply to unwatch :) 
dry-run is a bit down below, we skip ACL check currently, maybe we can move this down a bit.

File: src/raft.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/raft.c]: @fadidahanna ?

File: src/redisraft.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c]: prevent watch within MULTI.

File: src/redisraft.c [hunk 12]
```diff
@@ -555,6 +555,28 @@ static void handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
     RedisModule_ReplyWithSimpleString(ctx, "OK");
 }
 
+static void handleEndClientSession(RedisRaftCtx *rr, RaftReq *req, unsigned long long id, char *reason)
```
[the remark this answers is not in the document]
Comment (sjpotter) [reply src/redisraft.c:558]: ok applied

File: src/redisraft.c (hunk 11 above)
Comment (sjpotter) [reply src/redisraft.c:666]: it can.  so, we should.  moving.

File: src/multi.c [hunk 13]
```diff
@@ -103,6 +103,12 @@ bool MultiHandleCommand(RedisRaftCtx *rr,
 
     /* Are we in MULTI? */
     if (multiState->active) {
+        /* can't call WATCH within a MULTI, but doesn't error out MULTI */
+        if (cmd_len == 5 && !strncasecmp(cmd_str, "WATCH", 5)) {
+            RedisModule_ReplyWithError(ctx, "ERR WATCH inside MULTI is not allowed");
+            return true;
```
Comment (tezc) [inline src/multi.c:109]:
```suggestion
            multiState->error = true;
            return true;
```

Should we fail the transaction?

File: src/redisraft.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tezc) [reply src/redisraft.c]: pff, horrible. I hope we can reject unwatch as well

File: src/raft.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (fadidahanna) [reply src/raft.c]: We agreed in our daily to change raft session into session id, and then to solve the ambiguity, but we can do it later.
